Sources/KavaPlugin.swift

```diff
@@ -0,0 +1,251 @@
+    
+    public override class var pluginName: String {
+        return "KavaPlugin"
+    }
+    
+    public required init(player: Player, pluginConfig: Any?, messageBus: MessageBus) throws {
+        try super.init(player: player, pluginConfig: pluginConfig, messageBus: messageBus)
+        guard let config = pluginConfig as? KavaPluginConfig else {
+            PKLog.error("missing plugin config or wrong plugin class type")
+            throw PKPluginError.missingPluginConfig(pluginName: KavaPlugin.pluginName)
+        }
+        
+        self.config = config
+        self.registerEvents()
+    }
+    
+    public override func onUpdateMedia(mediaConfig: MediaConfig) {
+        super.onUpdateMedia(mediaConfig: mediaConfig)
+        self.resetPlayerFlags()
+        self.unregisterFromBoundaries()
+        self.stopViewTimer()
+    }
+    
+    public override func onUpdateConfig(pluginConfig: Any) {
+        super.onUpdateConfig(pluginConfig: pluginConfig)
+        
+        guard let config = pluginConfig as? KavaPluginConfig else {
+            PKLog.error("plugin config is wrong")
+            return
+        }
+        
+        PKLog.debug("new config::\(String(describing: config))")
+        self.config = config
+    }
+    
+    public override func destroy() {
+        self.messageBus?.removeObserver(self, events: playerEventsToRegister)
+        self.unregisterFromBoundaries()
+        self.stopViewTimer()
+        super.destroy()
+    }
+    
+    static func cleanPlaybackPoints() -> [KavaEventType : Bool] {
+        return playbackPoints.reduce([KavaEventType : Bool]()) { (dict, point) -> [KavaEventType : Bool] in
+            var dict = dict
+            dict[point] = false
+            return dict
+        }
+    }
+    
+    func registerToBoundaries() {
+        if let player = player, boundaryObservationToken == nil {
+            let boundaryFactory = PKBoundaryFactory(duration: player.duration)
+            let boundaries = playbackPoints.map({ boundaryFactory.percentageTimeBoundary(boundary: convertToPercentage(type: $0)) })
+            boundaryObservationToken = player.addBoundaryObserver(boundaries: boundaries, observeOn: nil) { [weak self] (time, percentage) in
+                self?.sendPercentageReachedEvent(percentage: Int(percentage * 100))
+            }
+        }
+    }
+    
+    func unregisterFromBoundaries() {
+        if let _ = boundaryObservationToken {
+            player?.removeBoundaryObserver(boundaryObservationToken!)
+            boundaryObservationToken = nil
+        }
+    }
+    
+    func setupViewTimer() {
+        if viewTimer == nil {
+            viewTimer = Timer.scheduledTimer(timeInterval: 10, target: self, selector: #selector(reportView), userInfo: nil, repeats: true)
+        }
+    }
+    
+    func stopViewTimer() {
+        viewTimer?.invalidate()
+        viewTimer = nil
+    }
+    
+    @objc func reportView() {
+        if let _ = bufferingStartTime {
```

Comment:
buffering can happen multiple times in a 10 seconds window for some cases, this should be an array of start + end time or something else that could handle multiple buffering events.